rules/opentitan_test.bzl

```diff
@@ -0,0 +1,374 @@
+        for dep in deplist:
+            uniq[dep] = True
+    return uniq.keys()
+
+def opentitan_functest(
+        name,
+        targets = ["dv", "verilator", "cw310"],
+        args = [],
+        data = [],
+        ottf = _OTTF_DEPS,
+        test_in_rom = False,
+        signed = False,
+        key = "test_key_0",
+        dv = None,
+        verilator = None,
+        cw310 = None,
+        **kwargs):
+    """A helper macro for generating OpenTitan functional tests.
+
+    This macro is mostly a wrapper around opentitan_flash_binary, but creates
+    testing artifacts for each of the hardware targets in `targets`. The testing
+    artifacts are then given to an `sh_test` rule which dispatches the test to
+    the corresponding hardware target via opentitantool.
+    Args:
+      @param name: The name of this rule.
+      @param targets: A list of hardware targets on which to dispatch tests.
+      @param args: Extra arguments (in addition to those defined in the target-
+                   specific parameter dictionary) to pass to the test runner
+                   (`opentitantool` or `dvsim.py`).
+      @param data: Extra data dependencies (in addition to those defined in the
+                   target-specific parameter dictionary) needed while executing
+                   the test.
+      @param ottf: Default dependencies for OTTF tests. Set to empty list if
+                   your test doesn't use the OTTF.
+      @param test_in_rom: Whether to run the test from ROM, runs from flash by
+                          default.
+      @param signed: Whether to sign the test image. Unsigned by default.
+      @param key: Which signed test image (by key) to use.
+      @param dv: DV test parameters.
+      @param verilator: Verilator test parameters.
+      @param cw310: CW310 test parameters.
+      @param **kwargs: Arguments to forward to `opentitan_flash_binary`.
+
+    This macro emits the following rules:
+        opentitan_flash_binary named: {name}_prog (and all emitted rules).
+        sh_test                named: dv_{name}
+        sh_test                named: verilator_{name}
+        sh_test                named: cw310_{name}
+        test_suite             named: {name}
+    """
+
+    # Generate flash artifacts for test.
+    deps = _unique_deps(kwargs.pop("deps", []), ottf)
+    if test_in_rom:
+        opentitan_rom_binary(
+            name = name + "_rom_prog",
+            deps = deps,
+            **kwargs
+        )
+    opentitan_flash_binary(
+        name = name + "_prog",
+        output_signed = signed,
+        deps = deps,
+        **kwargs
+    )
+
+    all_tests = []
+
+    target_params = {
+        "sim_dv": dv_params() if (dv == None) else dv,
+        "sim_verilator": verilator_params() if (verilator == None) else verilator,
+        "fpga_cw310": cw310_params() if (cw310 == None) else cw310,
+    }
+
+    for target, params in target_params.items():
+        if target.split("_")[-1] in targets:
```

Comment:
I would write this as `if not in: continue` to promote less nesting.

Reply:
Fixed.